packages/utils/src/test.js

```diff
@@ -0,0 +1,156 @@
+import * as exports from '.';
+
+describe('rowsFromArray', () => {
+  it('works as expected', () => {
+    const { rowsFromArray } = exports;
+    const fields = [
+      { key: 'one', heading: 'One' },
+      { key: 'two', heading: 'Two', transform: () => 'two' },
+      { key: 'three', heading: 'Three' },
+    ];
+    const array = [
+      {},
+      { one: 'test', two: 'test' },
+    ];
+    
+    let rows = rowsFromArray(array, fields, false);
+    expect(rows).toEqual([['-', 'two', '-'], ['test', 'two', '-']]);
+
+    rows = rowsFromArray(array, fields, true);
+    expect(rows).toEqual([['test', 'two', '-']]);
+  });
+});
+
+describe('rowsFromObject', () => {
+  it('works as expected', () => {
+    const { rowsFromObject } = exports;
+    const fields = [
+      { key: 'one', heading: 'One' },
+      { key: 'two', heading: 'Two', transform: () => 'hardcodedstring' },
+      { key: 'three', heading: 'Three' },
+    ];
+    const object = { one: 'test', two: 'test', three: null };    
+    const skipEmptyValues = true;
+    const rows = rowsFromObject(object, fields, skipEmptyValues);
+
+    expect(rows).toEqual({
+      rows: [
+        ['One', 'test'],
+        ['Two', 'hardcodedstring'],
+      ],
+      names: [
+        'one',
+        'two',
+      ],
+    });
+  });
+
+  it('can use a default transform', () => {
+    const { rowsFromObject } = exports;
+    const fields = [
+      { key: 'one', heading: 'One', names: ['firstCol', 'secondCol'] },
+      { key: 'two', heading: 'Two', transform: (val) => val },
+    ];
+    const object = { one: 'test', two: 'some val' };
+    const skipEmptyValues = true;
+    const deafultTransform = (david) => `${david} david`;
+    const rows = rowsFromObject(object, fields, skipEmptyValues, deafultTransform);
+
+    expect(rows).toEqual({
+      rows: [
+        ['One', 'test david'],
+        ['Two', 'some val'],
+      ],
+      names: [
+        ['firstCol', 'secondCol'],
+        'two',
+      ],
+    });
+  });
+
+  it('handles undefined values without skipping', () => {
+    const { rowsFromObject } = exports;
+    const fields = [
+      { key: 'one', heading: 'One' },
+      { key: 'two', heading: 'Two' },
+    ];
+    const object = { one: 'test', two: undefined };
+    const skipEmptyValues = false;
+    const rows = rowsFromObject(object, fields, skipEmptyValues);
+
+    expect(rows).toEqual({
+      rows: [
+        ['One', 'test'],
+        ['Two', ''],
+      ],
+      names: [
+        'one',
+        'two',
+      ],
+    });
+  });
+});
+
+describe('titlesFromFields', () => {
+  it('works as expected', () => {
+    const { titlesFromFields } = exports;
+    const fields = [
+      { key: 'one', heading: 'One' },
+      { key: 'two', title: 'Two' },
+    ];
+    const result = titlesFromFields(fields);
+    expect(result).toEqual(['One', 'Two']);
```

Comment:
on a general coding style point i prefer to see whitespace after variable declarations - one day i'll get around to setting a linting rule to enforce that 😜

Reply:
https://github.com/stevesims/govuk-frederic/issues/5 becomes ever more pressing! 😛